clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -117,6 +117,10 @@ public String id() {
         return id;
     }
 
+    public SelectionKey selectionKey() {
+        return transportLayer.selectionKey();
+    }
+
```

Comment:
I am not sure we want to expose selection key from `KafkaChannel` or `TransportLayer`. It makes it easy to use it or hold on to it. In `SelectorTest`, we have a method that uses reflection to check that all sets and maps in `Selector` are empty. What we are missing is a test with this scenario. Perhaps we could add a scenario in `SslSelectorTest` which extends `SelectorTest` and use `verifySelectorEmpty`?

Reply:
I was unsure about exposing the selection key as well. An alternative I considered is moving the key validity check to the top of the loop in `pollSelectionKeys`. I thought it was a little nicer keeping the invariant that we only process keys with attached channels, but perhaps having a null check is not a big deal. What do you think?

Ack on the testing suggestion.

Reply:
Hmm, I think null check could end up masking bugs. The alternative would be to store channels instead of keys and create a set with the keys when needed (a new empty set is created anyway in each `poll()` if the set is non-empty, we could just populate it in each poll loop). Or keep it as is, I will leave it up to you.

Reply:
Let me try that suggestion and see how it looks. I do prefer maintaining the channel objects instead of selection keys.

Reply:
I left it as is because we still need a way to find the selection key corresponding to the channel.